### ISSUE TITLE
when the smtp server auth with 'username', the smtp.py client will thow exception

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4055]
|<img alt="itfront's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| itfront reported|
|-|-|
|Trac ID|trac#4055|
|Type|defect|
|Created|2009-10-13 02:18:43Z|

when I use the smtp.py to send email,I got exception:

```
Unhandled Error
Traceback (most recent call last):
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/python/log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/python/log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/python/context.py", line 59, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/python/context.py", line 37, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/internet/pollreactor.py", line 186, in _doReadOrWrite
    why = selectable.doRead()
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/internet/tcp.py", line 463, in doRead
    return self.protocol.dataReceived(data)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/protocols/basic.py", line 231, in dataReceived
    why = self.lineReceived(line)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/mail/smtp.py", line 1052, in lineReceived
/13823106429
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/python/context.py", line 37, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/internet/pollreactor.py", line 186, in _doReadOrWrite
    why = selectable.doRead()
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/internet/tcp.py", line 463, in doRead
    return self.protocol.dataReceived(data)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/protocols/basic.py", line 231, in dataReceived
    why = self.lineReceived(line)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/mail/smtp.py", line 1052, in lineReceived
    why = self._okresponse(self.code,'\n'.join(self.resp))
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/mail/smtp.py", line 1372, in esmtpState_challenge
    self._authResponse(auth, resp)
  File "/usr/local/lib/python2.6/site-packages/Twisted-8.2.0-py2.6-linux-i686.egg/twisted/mail/smtp.py", line 1388, in _authResponse
    self.sendLine(encode_base64(resp, e
```

------------------------------------------------
finally , I find out that , at about line 1379:

```
def _authResponse(self, auth, challenge):
        self._failresponse = self.esmtpAUTHDeclined

        try:
            challenge = base64.decodestring(challenge)
            challenge = challenge.capitalize()  #add this!!
```

Attachments:

* [smtp.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/38f/38f6567cab76d3edc42f9366048b81a904f2e20d/92ce98af45617f7e7729494a2350cc67f59c68c6.py) (67549 bytes) - added by itfront on 2009-10-13 02:20:15Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4055 4055
type__defect defect
reporter__itfront itfront
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__mail mail
keywords__ 
time__1255400323000000 1255400323000000
changetime__1344948875000000 1344948875000000
version__None None
owner__itfront itfront

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="itfront's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|itfront commented|
|-|-|

I review the smtp.py and find out that:the error-input detection is too week..beacuse not all smtp server is coded good, you must tolerate non-standard!

**Comment by twisted-trac** [comment]
|<img alt="itfront's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|itfront set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to itfront|
|-|-|

Hi itfront.

Thanks for investigating this problem.

I wonder if you could provide your changes as a patch instead of just attaching a modified copy of smtp.py.  Patches are much easier to work with.

Also, it would be be really great if you could add a unit test which triggers this behavior.  Unit tests are a much better way to describe bugs than anything else.  They let us see what's really happening, reproduce the problem, and verify that a particular change fixes the problem.

Finally, when you've attached a patch with a unit test and a fix, please add the _review_ keyword to the ticket so we know it's ready to look at.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Fixing description markup

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_5"></a>|@MostAwesomeDude commented|
|-|-|

The suggested fix breaks test_crammd5Auth badly and I'm not sure how to fix it as the unit tests are somewhat indecipherable. :T Suggestions?

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_6"></a>|@argonemyth set status to `closed`|
|-|-|

If itfront means when an smtp server sends 'username' as a challenge, the smtp.py client will throw an exception, the problem is no longer there. Check the following `imap4.LOGINAuthenticator` code:


```
    def challengeUsername(self, secret, chal):
        # Respond to somethichallengesng like "Username:"
        self.challengeResponse = self.challengeSecret
        return self.user
```

It doesn't care about the value `chal` at all...The following simple test (not test case) should confirm this: 

In class `smtp.LOGINCredentials`, replacing 'Password:' and 'Username:' with whatever words you want, `test_smtp.AuthTestCase.test_loginAgainstWeirdServer` (which uses the `LOGINCredentials`) still passes.

I am closing this ticket because I think the bug is no longer there. If anyone could confirm that the bug is still present, feel free to re-open the ticket, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)|@argonemyth set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_8"></a>|@argonemyth set status to `closed`|
|-|-|

This was fixed in ticket [#10463](https://github.com/twisted/twisted/issues/10463) (r11917). Marking this one as a duplicate, since [#10463](https://github.com/twisted/twisted/issues/10463) is already resolved.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

[#10463](https://github.com/twisted/twisted/issues/10463) is somewhat related, but # seems more directly related.  [#10463](https://github.com/twisted/twisted/issues/10463) was resolved 5 years before this ticket was filed, and only changed code in `twisted/mail/imap4.py`.  [#4031](https://github.com/twisted/twisted/issues/4031) actually fixed the SMTP client by making it share that same code from the IMAP4 module.
